.NET/Microsoft.Recognizers.Text/ModelFactory.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace Microsoft.Recognizers.Text
+{
+    internal class ModelFactory<TModelOptions> : Dictionary<(string culture, Type modelType), Func<TModelOptions, IModel>>
+    {
+        private static ConcurrentDictionary<(string culture, Type modelType, string modelOptions), IModel> cache = new ConcurrentDictionary<(string culture, Type modelType, string modelOptions), IModel>();
+
+        public T GetModel<T>(string culture, TModelOptions options) where T : IModel
+        {
+            if (string.IsNullOrEmpty(culture))
+            {
+                throw new ArgumentNullException("culture", "Culture is required.");
+            }
+
+            // Look in cache
+            var cacheKey = (culture.ToLowerInvariant(), typeof(T), options.ToString());
+            if (cache.ContainsKey(cacheKey))
+            {
+                return (T)cache[cacheKey];
+            }
+
+            // Use Factory to create instance
+            var key = GenerateKey(culture, typeof(T));
+            if (this.ContainsKey(key))
+            {
+                var factoryMethod = this[key];
+                var model = (T)factoryMethod(options);
+
+                // Store in cache
+                cache[cacheKey] = model;
+
+                return model;
+            }
+
+            throw new ArgumentException($"Could not find Model with the specified configuration: {culture}, {typeof(T).ToString()}");
```

Comment:
The behaviour to fallback to a default culture was also removed, but it is required.
For example, in the PhoneNumber case, any culture should return the English model as it is a base model, non culture specific.
Also, in some other cases, if a culture doesn't have it's own model, it is desirable to fallback to English to at least get basic entity recognition. For example: trying to get a Russian date model, that doesn't exist, but still recognize "2018-12-13" as a date.

Reply:
We changed the `Recognizer<>` initialization to specify the `defaultCulture` (i.e. fallbackCulture) and moved the `culture` parameter to the `GetModel()` method. This way, if a given `culture` does not exists we fallback to the recognizer's default as shown in the following sample:
```C#
var recognizer = new DateTimeRecognizer(DefaultCulture);
Assert.AreEqual(recognizer.GetDateTimeModel(InvalidCulture), recognizer.GetDateTimeModel());
```